framework/src/csg/CSGUniverse.C

```diff
@@ -0,0 +1,98 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "CSGUniverse.h"
+
+namespace CSG
+{
+
+CSGUniverse::CSGUniverse(const std::string & name, bool is_root) : _name(name), _is_root(is_root) {}
+
+CSGUniverse::CSGUniverse(const std::string & name, std::vector<CSGCell *> & cells, bool is_root)
+  : _name(name), _is_root(is_root)
+{
+  for (auto cell : cells)
+    addCell(*cell);
+}
+
+void
+CSGUniverse::addCell(const CSGCell & cell)
+{
+  auto cell_name = cell.getName();
+  if (!hasCell(cell_name))
+    _cells.push_back(cell);
+  else
+    mooseWarning("Universe " + getName() + " already contains a cell by name " + cell_name + ". " +
+                 "Skipping cell insertion for cell with duplicate name.");
+}
+
+const CSGCell &
+CSGUniverse::getCell(const std::string & name)
+{
+  if (!hasCell(name))
+    mooseError("Cell with name " + name + " does not exist in universe " + _name + ".");
+  for (const CSGCell & cell : _cells)
+    if (cell.getName() == name)
```

Comment:
Is there a reason there isn't a name -> cell map? these linear searches could be expensive with enough cells